### ISSUE TITLE
Savestate issues resurfaced

### ISSUE THREAD

**Issue by fr500** [barbudreadmon/fbalpha2018#166]
NeoGeo Save states are no longer the same size between 32 and 64 bits making it impossible to netplay.
Haven't had a change to bisect.

CPS2 savestates work but music gets really messed up anytime a state is loaded. It only happens when it's between a 32 bit and a 64 bit host

**Comment by twinaphex** [comment]
@barbudreadmon Can we please avoid these issues from being perpetually reintroduced?

What happened in this instance? Did you do a rebase against upstream and did the faulty savestate code get re-merged again? Can you maybe maintain a patch file that you apply against upstream so this doesn't happen again?

Alternatively, maybe it's possible to get in contact with upstream and get them to merge this code so it doesn't happen again?

@fr500 is getting pretty frustrated by this and I can't blame him, these kind of core regressions are really quite bad considering there are a lot of people relying on FBA netplay to just work, and then all of a sudden it breaks horribly again. In this case, we at libretro have the ability to test on all these platforms and might therefore know best what is portable and what isn't , whereas upstream has the luxury of only having to care about Win32/Windows and therefore might not notice all these issues.

**Comment by fr500** [comment]
Well I'm not frustrated but users are
This is what CPS2 sounds like now after loading a state on netplay

https://photos.app.goo.gl/WWqK1ci0FFcVFTHP2

**Comment by barbudreadmon** [comment]
Every time there is a new release, i create a patch vanilla fba > libretro fba of previous release, then apply all changes to the new version. That's what i've been doing for 3 years now. I don't know why it's broken again. Actually i don't remember seeing any difference between vanilla .40 and libretro .40 for the neogeo driver. Cps2 differences should be there though.

**Comment by barbudreadmon** [comment]
There was a fix on the neogeo driver recently (a buggy speedhack was removed), perhaps that's where the issue comes from.

**Comment by barbudreadmon** [comment]
Now i remember, there was a fix for cross platform savestates in the m68k cpu driver, which would affect both cps2 and neogeo. It should still be there.

**Comment by furiadeoso** [comment]
Hi, @barbudreadmon . I'm testing this issue as user. Really the first time I've done crossplay between PC <-> Android was at the date : 12/03/2017 . I've found it in a discord log with my community team. 
Metal slug did't work in than moment. I'm not sure at all, but I think speedhacks fix was released two days after.
We can help you as testers if needed.
Best regards.

**Comment by barbudreadmon** [comment]
I think part of the fix was probably directly applied to the generated files in the m68k driver. Those files are regenerated with every release, that's part of the process when adding the new drivers. Those files are not present upstream (generated at build time) so i don't keep track of changes to them. I'll check when i'm back home. A proper fix would be to apply the changes to the code generating those.

**Comment by barbudreadmon** [comment]
It doesn't seem to be an issue with the generated files.
Are those issues already present in https://github.com/libretro/fbalpha/commit/bb8dce61562de05704fcdd88e7e3cc6e90d5a474 ?
I think i trashed some changes in the libretro-fbalpha code for that particular release (.40), because i previously submitted those changes (the ones for cross platform savestates compatibility) to upstream and they refactored those.

**Comment by fr500** [comment]
Well found a good commit which is something (testing CPS2 
Thankfully the issue does happen in x86 windows vs x86_64 windows so it's faster to bisect than windows vs android (still a pita to build so many commits in a row but whatever)

I'll let you know once I find the first bad commit

**Comment by fr500** [comment]
BTW seems CPS2 sound is always bad (but not like on the video) in windows x86

**Comment by fr500** [comment]
Found the culprits for CPS2

```
': not a valid identifierline 88: export: `dashless
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
bb8dce61562de05704fcdd88e7e3cc6e90d5a474
0d21221fe9030ca8c482ca9042f92babf0a52dde
We cannot bisect more!

```

So I guess upstream broke it

**Comment by fr500** [comment]
Same applies to CPS1/3/NeoGeo

**Comment by fr500** [comment]
Sucks, guess we need to report to upstream?

**Comment by barbudreadmon** [comment]
https://github.com/libretro/fbalpha/commit/bb8dce61562de05704fcdd88e7e3cc6e90d5a474 is the .40 release. It was supposed to be the first release upstream including this cross-platform compatibility for savestates. It could be that i did exactly what @twinaphex said (submitting those changes upstream) and it backfired when i trusted their refactoring of this code.

In https://github.com/libretro/fbalpha/commit/0d21221fe9030ca8c482ca9042f92babf0a52dde i see @aliaspider did a change in m68kcpu.c to "m68k_context_size_no_pointers" to fix warnings/errors (there is not much precision about this). This function in particular is "Used to calculate the context size minus the system-specific pointers, for savestates", so it could be the perfect culprit. It would be interesting to see if the bug is present if you switch back from "intptr_t" to "int"

**Comment by barbudreadmon** [comment]
I'm also wondering if https://github.com/libretro/fbalpha/commit/5f7f2258214300c886fa422497be3a11ea0c462e is not a fix which is the result of the changes to m68k in https://github.com/libretro/fbalpha/commit/0d21221fe9030ca8c482ca9042f92babf0a52dde . I would be curious to see if changing back to ACB_VOLATILE in libretro.cpp wouldn't be another part of the fix if changing back to "int" in m68kcpu.c is not enough.

**Comment by fr500** [comment]
No,  without FULLSCAN there were lots of determinism issues (for instance
the number of credits didn't save in neogeo)

I think they may have fudged their fix (by not testing)

**Comment by barbudreadmon** [comment]
Ok, did you try reverting intptr_t to int in "m68k_context_size_no_pointers" ? Perhaps that's not the only problem here but it's definitely part of the issue (it will produce different results on 32/64 bits).

**Comment by fr500** [comment]
No, not yet, I'm at work atm

**Comment by fr500** [comment]
The FIX Savestates again commit didn't fix padding / size. It fixed incomplete states, basically I tested all the saving options to find out what do I need.

FULLSCAN is proper for complete reliable savestates.
Ali's commit was about build errors we were experiencing. I can try reverting that one this weekend.

**Comment by fr500** [comment]
Actually give me a few gonna try right now seems simple enough

**Comment by fr500** [comment]
```[ERROR] CMD_LOAD_SAVESTATE received an unexpected save state size.```

Nope sizes still differ

**Comment by barbudreadmon** [comment]
Ok, i'll take a look this week-end or next week.

**Comment by barbudreadmon** [comment]
Did a quick compare of what you did prior to the .40 bump with what was included in .40 upstream, there were missing parts, i added them.
There is just this part i don't know how to deal with : https://pastebin.com/0cmkDAAi : the actual code (line 3500-3505 of https://github.com/libretro/fbalpha/tree/master/src/cpu/sh2) is slightly different from what it used to be : 
The old code in .39
`ScanVar(& ( Sh2Ext[i].sh2 ), sizeof(SH2) - 4, szText);`
was replaced by this in current version
`ScanVar(& ( Sh2Ext[i].sh2 ), sizeof(SH2), szText);`
`Sh2Ext[i].sh2.irq_callback = irq_callback;`

**Comment by fr500** [comment]
Ahh, well that was all ali, I can try to ask him next time I see him around, thanks for taking a look.

**Comment by fr500** [comment]
NeoGeo and CPS2 are fixed again, CPS3 isn't yet.
I checked the neo source thread, if you want I can try to log the sizes of the variables that go into the saves, that's the analysis I performed the first time anyhow.

Not sure if it's worth mentioning or not, but afaik, long differs between linux and windows for instance.

**Comment by barbudreadmon** [comment]
I'll take a look at sh2.cpp this week-end. I should be able to test this by myself if i use the desktop + raspberry.

**Comment by barbudreadmon** [comment]
INT32 is actually a short, so it should have a size/alignment of 4 with any compiler/arch. In my commit i think the ALIGN_VAR(8) before the INT32 are indeed not needed, the only one needed is probably the one before the void in ay8910.h

**Comment by barbudreadmon** [comment]
https://github.com/libretro/fbalpha/commit/f34cde4478216d59f7bcaec1075f188b27c61b4c seems to solve the issue in sh2 somehow, the new code looks horrible, and for whatever reason i can't get the stages background to load properly after restarting retroarch if i don't remove your hack in cps3run.cpp (that's one of the hacks for savestates you introduced in .39), which ends up creating quite big savestates.

**Comment by fr500** [comment]
heh the stages thing is... well on purpose.
It's a **HUGE** speedboost for netplay / rewind, maybe we could make it optional in the future.

I'm actually thinking about a netplay awareness environment call, we could make the hack optional for netplay.

**Comment by fr500** [comment]
The stages should load after the next stage loads right?, just got up, gonna test after breakfast, thanks for taking time to check this out.

**Comment by furiadeoso** [comment]
Warn if you need me to do any tests :)

**Comment by barbudreadmon** [comment]
Yes, being able to differentiate between a normal savestate loading and a loading for netplay purpose seems like a great idea, not having the background when this is a normal savestate loading seems really bad to me.
I'll do some cleanup on the fixes since some of them are not needed

**Comment by barbudreadmon** [comment]
I also added back the no background loading hack for now.

**Comment by fr500** [comment]
Actually, CPS3 is good, but neogeo and CPS2 are fudged again.

**Comment by barbudreadmon** [comment]
Doesn't make much sense, size/alignment of INT32 (which is an int) should be the same on 32/64 bits. I'll investigate and add back the align_var if i can't find another workaround.

**Comment by fr500** [comment]
MINGW32 and MINGW64, the easiest test case is Windows x86 vs Windows x86_64 since I can run both in the same box at the same time.

**Comment by fr500** [comment]
```Using built-in specs.
COLLECT_GCC=E:/Tools/msys64/mingw64/bin/gcc
COLLECT_LTO_WRAPPER=E:/Tools/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/6.3.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-6.3.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/x86_64-w64-mingw32/include --libexecdir=/mingw64/lib --enable-bootstrap --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,objc,obj-c++,fortran,ada --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-time=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev3, Built by MSYS2 project' --with-bugurl=https://sourceforge.net/projects/msys2 --with-gnu-as --with-gnu-ld
Thread model: posix
gcc version 6.3.0 (Rev3, Built by MSYS2 project)
```

```
Using built-in specs.
COLLECT_GCC=E:/Tools/msys64/mingw64/bin/gcc
COLLECT_LTO_WRAPPER=E:/Tools/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/6.3.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-6.3.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/x86_64-w64-mingw32/include --libexecdir=/mingw64/lib --enable-bootstrap --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,objc,obj-c++,fortran,ada --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-time=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev3, Built by MSYS2 project' --with-bugurl=https://sourceforge.net/projects/msys2 --with-gnu-as --with-gnu-ld
Thread model: posix
gcc version 6.3.0 (Rev3, Built by MSYS2 project)

```

**Comment by barbudreadmon** [comment]
Added back the dirty fix since i won't be able to fix this right away, the issue seems to be with "AY8910 #0" which is 264 byte long in 32 bits and 272 in 64 bits, which is weird, i'll try to get to the bottom of this, especially since it could be related to the sound issues in #155 and #163

**Comment by fr500** [comment]
yep it's fine now, but you're right the sound is pretty bad in 32 bits x86

**Comment by barbudreadmon** [comment]
What about this last commit ? I get the same sizes on arm32/x86_64, the savestates seems portable and the sound seems fine.

**Comment by fr500** [comment]
Ok let me rephrase the sound is bad is not savestate related, it always is, there is a permanent BZZZZZ in 32 bit x86 for some reason, the latest commit reintroduces the sound artifacts in CPS2 though. NeoGeo is fine.

**Comment by barbudreadmon** [comment]
Ok, i reverted this. This is really weird, there is probably something i don't understand about alignment, i'll try to find some good documentation about this.

**Comment by barbudreadmon** [comment]
Ok, i had a misunderstanding of what alignment does, so after reading about it, i think i understand why and where it is needed, an example : 

> struct AY8910
{
	INT32 Channel; // int32 (4)
	INT32 SampleRate; // int32 (4)
// Next
	read8_handler PortAread; // void (4) or (8) depending on arch
// next variable won't fit in the current 8 byte if arch is 64 bits, we need to force a new 8 bytes boundary
	ALIGN_VAR(8) read8_handler PortBread; // void (4) or (8) depending on arch
// next variable won't fit in the current 8 byte if arch is 64 bits, we need to force a new 8 bytes boundary
	ALIGN_VAR(8) write8_handler PortAwrite; // void (4) or (8) depending on arch
// next variable won't fit in the current 8 byte if arch is 64 bits, we need to force a new 8 bytes boundary
	ALIGN_VAR(8) write8_handler PortBwrite; // void (4) or (8) depending on arch
// next variable won't fit in the current 8 byte if arch is 64 bits, we need to force a new 8 bytes boundary
	ALIGN_VAR(8) INT32 register_latch; // int (4)
// Next (2 boundaries)
	UINT8 Regs[16]; // char (1) * 16
// Next
	INT32 lastEnable; // int (4)
	UINT32 UpdateStep; // uint (4)
// Next (5 boundaries)
	INT32 PeriodA,PeriodB,PeriodC,PeriodN,PeriodE; // int (4) * 5
	INT32 CountA,CountB,CountC,CountN,CountE; // int (4) * 5
// Next (2 boundaries)
	UINT32 VolA,VolB,VolC,VolE; // int (4) * 4
// Next
	UINT8 EnvelopeA,EnvelopeB,EnvelopeC; // uint8 (1) * 3
	UINT8 OutputA,OutputB,OutputC,OutputN;  // uint8 (1) * 4
	INT8 CountEnv; // int8 (1)
// Next
	UINT8 Hold,Alternate,Attack,Holding;
	INT32 RNG;
// Next (16 boundaries)
	UINT32 VolTable[32];
};

Basically it acts more like a "pad the previous variable, because the next one won't fit in current boundary in some cases" thing.
So yes, the align_var on register_latch is needed, it doesn't seem the one on PortRead is needed though

**Comment by barbudreadmon** [comment]
Some good news : next release of fbalpha got the pointers at the end of the struct, and use offsetof to determine which part of the struct need to be dumped. It should help a lot with the cross compatibility of the savestates, and should also make them slightly smaller.

**Comment by fr500** [comment]
nice! great news!

**Comment by furiadeoso** [comment]
Hi! Which is the driver of "in the hunt" (irem)? I believe it could have the same problem. Yesterday it was impossible to connect netplay between windows 10 and android.

**Comment by barbudreadmon** [comment]
I didn't commit this since i'm not sure it won't break anything, but here is the patch to remove all pointers from current code : https://pastebin.com/76Z8YYjt . If someone want to test it...

**Comment by barbudreadmon** [comment]
@furiadeoso @fr500 FYI, i applied the changes to remove pointers from savestates, let me know if it solved the "in the hunt" issue or if some new issue appeared.

**Comment by furiadeoso** [comment]
@barbudreadmon These changes have broken totally netplay between 64 and 32 bits. We have tested it using DDSOM, PUNISHER, and DINO, . Between Windows 10 64 bits and Android or Raspberry Pi. Before the changes all worked perfectly in crossplay. After that it's impossible to connect.
Rasp + android connection remains ok. So I understand something is worse with savestates now.

**Comment by barbudreadmon** [comment]
@furiadeoso did you try any system besides cps1/cps2 ?

**Comment by barbudreadmon** [comment]
I reported this to @dinkc64, seems like the QChan # differ in size between 32 and 64 bits

**Comment by furiadeoso** [comment]
@barbudreadmon I tried irem m92 ("in the hunt") too. But no more. Do you want me to try another?

**Comment by furiadeoso** [comment]
Aliens (konami) is not working too.
Nevertheless, Brapboys (kaneko- but I don't know the driver)  is working ok between 64-32. I'll report more later if neccessary.

**Comment by barbudreadmon** [comment]
This is enough for now, we first need to find out why the size differs with those :
- cps1/2 : "QChan #" 52 / 56
- m92 : "All Ram" 155760 / 155784
- aliens : "KonamiCPU Registers" 52 / 56

I checked the code for "QChan" and "KonamiCPU Registers", they both use offsetof on a pointer free struct, so it seems there is something we are overlooking.

**Comment by barbudreadmon** [comment]
My guess is that the pointer don't start at the same place between 32 and 64 bits because of alignment, and since we are dumping everything before the start of the pointer (by using the pointer's offset), the size differ.
We should probably define some macro similar to this : 
`#define STRUCT_SIZE_TO_MEMBER(type, member) offsetof(struct type, member) + sizeof(((struct type*)0)->member)`
use the struct name as first parameter, and the name of the member BEFORE the pointer as second parameter.

**Comment by barbudreadmon** [comment]
Let me know if my last commit fixes the issue on cps and aliens (m92 seems to be unrelated to the offsetof issue, i'll investigate it later)

**Comment by furiadeoso** [comment]
@barbudreadmon I've just try yesterday commit. Cps and aliens now is connecting.
I'm observing some kind of random"fast forward effect", but I don't know if it can be related to these last changes in the core.

**Comment by barbudreadmon** [comment]
Hmmm i don't think there should be any side effect, those changes should only remove unneeded stuff from the savestates, making them smaller and cross platform.

About irem drivers, m92 and m107 should be fixed, m90 doesn't support savestates at the moment (should be included in a future release), m62/m63/m72/vigilant already seemed fine.

**Comment by barbudreadmon** [comment]
@furiadeoso Also, if you find another driver with savestates issues, let me know !

**Comment by furiadeoso** [comment]
Thanks, @barbudreadmon . "In the hunt" is Ok. Only it happens as with CPS3: the first stage appears witch a black background. But after a reset or new stage is all ok.
I will continue reporting if I find new savestates issues.

**Comment by fr500** [comment]
CPS3 doesn't save sprite data to memory to speed up netplay.

**Comment by dinkc64** [comment]
maybe one could check if netplay is enabled in the driver, if not, scan the data needed for proper local savestates?

**Comment by fr500** [comment]
That's not currently possible, there is no way for the core to figure that
out but I have been working on a proposal for netplay awareness (for other
reasons) maybe then it can be added

**Comment by barbudreadmon** [comment]
> but I have been working on a proposal for netplay awareness (for other reasons) maybe then it can be added

Yeah, we talked about it in the commit about cps3's savestates, let me know as soon as this is available, i feel bad for the people who use savestates outside of netplay...

m92 doesn't save sprite data either ?

**Comment by furiadeoso** [comment]
I only tried "in the hunt" from m92, but I'm going to try undercover cops and so, and report again.
Is there a list with the drivers that support savestates? I could continue testing all of them.

**Comment by barbudreadmon** [comment]
> Is there a list with the drivers that support savestates

Most of them do, but sadly i don't think there is a list, actually i don't think dink noticed m90 was missing this feature before i mentioned it.

**Comment by furiadeoso** [comment]
Ok. Humm. And.. A list of arcade drivers? Some way to systematically we can check and discard all the drivers.

**Comment by furiadeoso** [comment]
@barbudreadmon We are trying to solve a fundamental problem for netplay with fba, and that could be related to the way in which savestates are created.
Could you take a look at the latest posts from https://github.com/libretro/RetroArch/issues/6309?
In particular there are some comments from @GregorR that could be relevant here:

`fbalpha's savestate code is nondeterministic independently of the actual behavior of the core. That is, you can save the exact same frame twice and result in two different savestates. As a result, the desynchronization detection also doesn't work.`
`Given that fbalpha seems perhaps not deeply nondeterministic, if they at least made the savestate code more reliable, check_frames could be sufficient to make it usable.`
`Probably a timestamp or some global counter not relevant to execution or something like that is getting saved. Or, maybe the states are compressed in a nondeterministic way or something like that. The states work fine, just I can call core_serialize twice and get different ones.`
Thanks!

**Comment by fr500** [comment]
I think this is something that should be brought up to upstream.

**Comment by barbudreadmon** [comment]
Do you have any lead to what could be done to fix it ? There is no netplay upstream and that's an issue that only affects netplay if i understand well, dink was nice enough to depointerize the savestates but i don't want to ask him more fix targeted only at retroarch's features (NB : to my knowledge, fightcade's netplay is not suffering from this).

**Comment by JordiRos** [comment]
Hi, I was talking to furiadeoso, and he told me about this issue. As far as I know, FBA is deterministic, but some drivers do not save all variables correctly because they are not "necessary", as in, they usually work in a save/load environment on a single machine, and it's not important if state changes a bit.
Original GGPO code (from Ponder) had fixes for exactly these situations in some drivers (CPS2, CPS3, NeoGeo, and a few others), because it relies on determinism for doing checks and rollbacks. The same is happenning in the new MegaDrive driver, but I haven't had time to look into that because that is new...
I could do a branch and build RA with these changes, if you think it is worth it? We could start with CPS2 driver, it's a good test bed.

**Comment by GregorR** [comment]
GGPO and RetroArch use the same underlying netplay technique. Anything that's needed for GGPO to do netplay successfully is also needed by RetroArch. So, if those changes are just lying around, it would certainly be appreciated :)

**Comment by GregorR** [comment]
Oh, and I should add: It hadn't occurred to me that the savestates could be nondeterministic in this particular way (they're merely missing some data), so indeed, my original assumption that the core itself is nondeterministic is probably wrong, as my only way of measuring that is by transferring a savestate and then following the same inputs. I suspect the core is fine, but just needs some tidbits added to savestates.

**Comment by JordiRos** [comment]
Sounds like is worth testing then. I will try to get some time this week for this

**Comment by fr500** [comment]
In any case, you wouldn't need to fork / branch retroarch, but fba libretro :)
Glad someone is looking into it now

**Comment by dinkc64** [comment]
I'll try to help where I can..

I did a quick test with savestates on the official/Windows version, doing 2 states in succession actually does return the same info.  The only requirement for the savestate staying exactly the same is that the frame doesn't advance between states (aka paused):
C:\mingw\msys\1.0\home\user\fbadebugg\savestates>fc /b "md_megaman slot 01.fs.backup0" "md_megaman slot 01.fs"
Comparing files md_megaman slot 01.fs.backup0 and MD_MEGAMAN SLOT 01.FS
FC: no differences encountered

**Comment by dinkc64** [comment]
JordiRos, Would you share the fixes that you have?
I've been working hard on improving the MD emulation, so it is possible that with the new core update something had been overlooked.

**Comment by JordiRos** [comment]
I am on it, I will let you know when I do the PR

**Comment by dinkc64** [comment]
thanks :)

**Comment by JordiRos** [comment]
https://github.com/JordiRos/fbalpha/tree/fix/sync

@furiadeoso is still testing the changes. I ignore how the netcode in RA works, but it seems Ponder presumably made these changes in GGPO to fix sync issues, and with those changes GGPO (and now FC) never gets out of sync. If they still happen in RA maybe there's something different in how network is implemented, although from what I know it should be the same (so maybe it's an undetected bug?). Problem is that these kind of issues are very hard to detect, as GregoR said. But if GGPO doesn't ever desync with this code, it should also be possible in RA.
I am not the author of these changes, but I can try to help answering if there's any doubt.

**Comment by barbudreadmon** [comment]
I guess we really need an environment variable for netplay.

**Comment by dinkc64** [comment]
regarding neo_upd4990a.cpp, this is an invalid date and could in theory cause problems (probably nothing serious).
find: uPD4990A.nYear = 2008;
replace w/: uPD4990A.nYear = 2008 % 100;

**Comment by JordiRos** [comment]
Thanks dink! your comments are great and very useful indeed

**Comment by furiadeoso** [comment]
I have been doing local netplay tests, 64 bits, with cps1 and cps2 using the original changes of Ponder. And comparing it with a core without these changes.

The main thing is to note that I can confirm that the serious problems of desynchronization were due to an RA bug, which affects only the stable version 1.7.1. It has already been fixed in a nightly version.

After this correction of RA, both old versions of the core and the one modified by @JordiRos  (Ponder mods) behave apparently the same, without desynchronizing.

The next step to test the netplay would be to do it between different machines and architectures, and incorporating the corrections of @dinkc64. @barbudreadmon , Are you going to incorporate those changes to this libretro core soon or do you want to facilitate them in some way to be tested before?

Thanks!

**Comment by barbudreadmon** [comment]
dink sent me lots of things today, including what is mentioned here, i intend to work on it this week-end.

**Comment by dinkc64** [comment]
furiadeoso: so it was an ra bug after all. :)  That's what I figured.. but, these changes from JordiRos/Ponder will definitely strengthen our state support even if they don't seem significant right now.

thanks guys,
dink

**Comment by furiadeoso** [comment]
Indeed, @dinkc64 :) . Still, in certain situations (infrequent and not yet well specified) the netplay can desynchronize without doing a check for a full savestate recovery (however fightcade does not need it). And there are also systems or games that are worse than others in the netplay. So I hope that @JordiRos /Ponder changes may be able to help improve these aspects.

Thanks to you,

**Comment by barbudreadmon** [comment]
I added the changes, i also reworked the netplay hack in cps3 to use the same kNetGame global variable introduced in the new neogeo rtc code. For now, this variable is always at 1, again i insist : we need something in the libretro api to differentiate between normal savestates and netplay savestates.

**Comment by GregorR** [comment]
I also want to get a netplay API in, and allow cores to know when they're in netplay. My reasons are very slightly different (I'd like for cores to be able to save state “differentially”, only changing what's actually changed, to hopefully accelerate saving), but a similar API is needed. At some point I'll be musing on the best way to introduce such an API.

**Comment by furiadeoso** [comment]
Thanks to all of you!! . I'm going to start testing today in my community with the core changes.

**Comment by barbudreadmon** [comment]
Closing this one too, supposedly there is no issue with netplay atm
